From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: BuildTools <unconfigured@null.spigotmc.org>
Date: Wed, 4 Jan 2023 12:14:16 +0800
Subject: [PATCH] Change branch


diff --git a/src/main/java/com/destroystokyo/paper/PaperVersionFetcher.java b/src/main/java/com/destroystokyo/paper/PaperVersionFetcher.java
index 9ba99bc5814a74dadf33034660f4dfc9cd530b94..4d5724360ec98bd005a0606821e3cb5d44fc330d 100644
--- a/src/main/java/com/destroystokyo/paper/PaperVersionFetcher.java
+++ b/src/main/java/com/destroystokyo/paper/PaperVersionFetcher.java
@@ -34,7 +34,7 @@ public class PaperVersionFetcher implements VersionFetcher {
     @Override
     public Component getVersionMessage(@Nonnull String serverVersion) {
         String[] parts = serverVersion.substring("git-Hearse-".length()).split("[-\\s]");
-        final Component updateMessage = getUpdateStatusMessage("NaturalCodeClub/Hearse", "master/" + getMinecraftVersion(), parts[0]);
+        final Component updateMessage = getUpdateStatusMessage("NaturalCodeClub/Hearse", "main/" + getMinecraftVersion(), parts[0]);
         final Component history = getHistory();
 
         return history != null ? Component.join(net.kyori.adventure.text.JoinConfiguration.separator(Component.newline()), history, updateMessage) : updateMessage;
