From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: wangxyper <wangxyper@163.com>
Date: Wed, 4 Jan 2023 19:26:47 +0800
Subject: [PATCH] Fix and change something


diff --git a/src/main/java/net/minecraft/server/level/ServerLevel.java b/src/main/java/net/minecraft/server/level/ServerLevel.java
index 0bc7eee8fe74697ced3e65ea33fc373f81d16d2f..1383031f6525cc0b18775fbd48d324c3ff12f05f 100644
--- a/src/main/java/net/minecraft/server/level/ServerLevel.java
+++ b/src/main/java/net/minecraft/server/level/ServerLevel.java
@@ -1731,8 +1731,9 @@ public class ServerLevel extends Level implements WorldGenLevel {
     @Override
     public void sendBlockUpdated(BlockPos pos, BlockState oldState, BlockState newState, int flags) {
         if (this.isUpdatingNavigations) {
-            String s = "recursive call to sendBlockUpdated";
-
+            if (Thread.currentThread() instanceof AnotherTickThread){
+                return;
+            }
             Util.logAndPauseIfInIde("recursive call to sendBlockUpdated", new IllegalStateException("recursive call to sendBlockUpdated"));
         }
 
diff --git a/src/main/java/net/minecraft/world/entity/Entity.java b/src/main/java/net/minecraft/world/entity/Entity.java
index 35afff85bcd4019789e4c40833b35df15a373b8c..230d9a08049c3c7bfa61c38844f3da26b260dce0 100644
--- a/src/main/java/net/minecraft/world/entity/Entity.java
+++ b/src/main/java/net/minecraft/world/entity/Entity.java
@@ -4453,7 +4453,7 @@ public abstract class Entity implements Nameable, EntityAccess, CommandSource {
         // Paper end
         // Paper start - rewrite chunk system
         if (this.updatingSectionStatus) {
-            LOGGER.error("Refusing to update position for entity " + this + " to position " + new Vec3(x, y, z) + " since it is processing a section status update", new Throwable());
+            LOGGER.error("Refusing to update position for entity " + this + " to position " + new Vec3(x, y, z) + " since it is processing a section status update");
             return;
         }
         // Paper end - rewrite chunk system
