From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: wangxyper <wangxyper@163.com>
Date: Thu, 5 Jan 2023 07:52:02 +0800
Subject: [PATCH] Fix a NPE in LivingEntity


diff --git a/src/main/java/net/minecraft/world/entity/LivingEntity.java b/src/main/java/net/minecraft/world/entity/LivingEntity.java
index 62eb1a3c209a5489b04765ff27492ad6912cc737..ebeefe4b7aa3e33f3df8dd4113206acdb94050d5 100644
--- a/src/main/java/net/minecraft/world/entity/LivingEntity.java
+++ b/src/main/java/net/minecraft/world/entity/LivingEntity.java
@@ -2014,15 +2014,18 @@ public abstract class LivingEntity extends Entity {
             BlockPos blockposition = this.blockPosition();
             BlockState iblockdata = this.getFeetBlockState();
 
-            if (iblockdata.is(BlockTags.CLIMBABLE)) {
-                this.lastClimbablePos = Optional.of(blockposition);
-                return true;
-            } else if (iblockdata.getBlock() instanceof TrapDoorBlock && this.trapdoorUsableAsLadder(blockposition, iblockdata)) {
-                this.lastClimbablePos = Optional.of(blockposition);
-                return true;
-            } else {
-                return false;
+            if (iblockdata!=null){
+                if (iblockdata.is(BlockTags.CLIMBABLE)) {
+                    this.lastClimbablePos = Optional.of(blockposition);
+                    return true;
+                } else if (iblockdata.getBlock() instanceof TrapDoorBlock && this.trapdoorUsableAsLadder(blockposition, iblockdata)) {
+                    this.lastClimbablePos = Optional.of(blockposition);
+                    return true;
+                } else {
+                    return false;
+                }
             }
+            return false;
         }
     }
 
