From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: wangxyper <wangxyper@163.com>
Date: Fri, 6 Jan 2023 08:17:11 +0800
Subject: [PATCH] Fix my careless error


diff --git a/src/main/java/net/minecraft/server/MinecraftServer.java b/src/main/java/net/minecraft/server/MinecraftServer.java
index 32807652eff529bec1a680721353fd61cf94edb7..9356b894ddef2683a4e0bbd7461896dc27e6ecd1 100644
--- a/src/main/java/net/minecraft/server/MinecraftServer.java
+++ b/src/main/java/net/minecraft/server/MinecraftServer.java
@@ -947,6 +947,7 @@ public abstract class MinecraftServer extends ReentrantBlockableEventLoop<TickTa
     // CraftBukkit end
 
     public void stopServer() {
+        HearseConfig.save();//Hearse
         // CraftBukkit start - prevent double stopping on multiple threads
         synchronized(this.stopLock) {
             if (this.hasStopped) return;
diff --git a/src/main/java/net/minecraft/world/ticks/LevelTicks.java b/src/main/java/net/minecraft/world/ticks/LevelTicks.java
index 5711d6d87a502978a6be1dd5984ee587f0633e9d..5bb00291d22a73dd78f18b904d6a68a94743644c 100644
--- a/src/main/java/net/minecraft/world/ticks/LevelTicks.java
+++ b/src/main/java/net/minecraft/world/ticks/LevelTicks.java
@@ -212,10 +212,10 @@ public class LevelTicks<T> implements LevelTickAccess<T> {
     private void cleanupAfterTick() {
         if (this.allRunned){
             this.toRunThisTick.clear();
+            this.containersToTick.clear();
+            this.toRunThisTickSet.clear();
         }
-        this.containersToTick.clear();
         this.alreadyRunThisTick.clear();
-        this.toRunThisTickSet.clear();
     }
 
     @Override
