From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: wangxyper <wangxyper@163.com>
Date: Fri, 6 Jan 2023 16:35:40 +0800
Subject: [PATCH] Fix a stupid error


diff --git a/src/main/java/net/minecraft/server/MinecraftServer.java b/src/main/java/net/minecraft/server/MinecraftServer.java
index aaaea3d2dc12f33bfa437845d2b933cf75b7b74f..aeff5dd89a6bb05cabd469c38242b7b136a10e6f 100644
--- a/src/main/java/net/minecraft/server/MinecraftServer.java
+++ b/src/main/java/net/minecraft/server/MinecraftServer.java
@@ -286,7 +286,6 @@ public abstract class MinecraftServer extends ReentrantBlockableEventLoop<TickTa
     protected WorldData worldData;
     private volatile boolean isSaving;
 
-    public boolean asyncEntityMoveEnabled;
     public boolean asyncEntityEnabled;
     private final boolean asyncExecutorEnabled;
     private int asyncExecutorThreadCount;
@@ -425,7 +424,7 @@ public abstract class MinecraftServer extends ReentrantBlockableEventLoop<TickTa
         HearseConfig.init();
         this.asyncEntityEnabled = HearseConfig.getBoolean("enable-async-entity",true);
         this.asyncExecutorThreadCount = HearseConfig.getInt("async-entity-executor-thread-count",Runtime.getRuntime().availableProcessors());
-        this.asyncExecutorEnabled = this.asyncEntityMoveEnabled;
+        this.asyncExecutorEnabled = this.asyncEntityEnabled;
         if (this.asyncExecutorEnabled){
             this.asyncExecutor = new CallbackExecutor(
                     this.asyncExecutorThreadCount,
